### PR TITLE
perf(ucp): batch variation fetching for /catalog/search and /catalog/lookup

### LINKED ISSUE #349: UCP 2026-04-08 shape compliance pass (0.12.0)
## Summary

Audit on live `/catalog/lookup` for products 22 (simple) and 35 (variable, "T-Shirt with Logo") found 6 critical + 5 important gaps against the canonical UCP `2026-04-08` schemas. All gaps are mechanical shape changes — no new endpoints needed, no behavior changes.

**Compliance, not breaking.** A spec-conforming UCP agent reading our current responses today is already getting rejected by `variant.json` validation (missing required `price`) — we ship `PROTOCOL_VERSION = '2026-04-08'` in our manifest but emit non-conformant payloads. This PR aligns the implementation with the spec we already declare. The only consumer that would experience this as "breaking" is one bypassing spec validation entirely and reading our specific non-conformant keys; we have no such public integrations. Version bump 0.11.x → 0.12.0 reflects the substantive wire-format scope, not a contract break.

Spec verified directly against `https://github.com/Universal-Commerce-Protocol/ucp` `release/2026-04-08` (HEAD `6373733`).

### 1. Variant `price` field rename + `list_price` semantics flip

Spec (`variant.json`):
- `required: ["id", "title", "description", "price"]`
- `price` (required) — "Current selling price." ($ref `price.json`)
- `list_price` (optional) — "List price before discounts (for strikethrough display)."

Today we emit:
- `list_price` carrying the current price (wrong field name)
- `compare_at_price` carrying the strikethrough (non-spec)

Fix: rename `list_price` → `price`, rename `compare_at_price` → `list_price`.

Files:
- `includes/ai-storefront/ucp-rest/class-wc-ai-storefront-ucp-variant-translator.php:69, :91` — `translate()` path
- Same file `:179, :185` — `synthesize_default()` path

### 2. `selected_option.json` shape

Spec: `required: ["name", "label"]`; optional `id`.

Today: `[{attribute, value}]`. Fix: rename `attribute` → `name`, `value` → `label`.

Files:
- `class-wc-ai-storefront-ucp-variant-translator.php:559-562` — array-shape path
- Same file `:574-578` — parsed-string path

### 3. `option_value.json` shape on `product.options[].values[]`

Spec (`product_option.json` + `option_value.json`):
- `product_option.values[]` items reference `option_value.json`
- `option_value: required: ["label"]`, optional `id`

Today: bare strings `["Black", "Green", "Navy"]`. Fix: wrap each as `[{label: "Black"}, ...]`.

Files:
- `class-wc-ai-storefront-ucp-product-translator.php:746-758` — `extract_classified_attributes()`

### 4. Per-variant `inputs[]` correlation in `/catalog/lookup`

Spec (`catalog_lookup.json#/$defs/lookup_variant`):
```json
{
  "allOf": [
    { "$ref": "types/variant.json" },
    {
      "required": ["inputs"],
      "properties": {
        "inputs": {
          "type": "array",
          "items": { "$ref": "types/input_correlation.json" },
          "minItems": 1
        }
      }
    }
  ]
}
```

Today: top-level `inputs: ["prod_22","prod_35"]` echo on the envelope. Fix: drop the envelope echo, attach `inputs: [{id, match}]` to each emitted variant per `input_correlation.json`. `match` enum: `"exact"` (variant ID requested directly) or `"featured"` (product ID resolved to a featured variant).

Files:
- `class-wc-ai-storefront-ucp-rest-controller.php:1700` — top-level `inputs` emission
- New: per-variant attachment during translate-loop

### 5. Lookup-response definition path

Spec lives at `catalog_lookup.json#/$defs/lookup_response` (not `responses/lookup.json`). No code change — affects test fixtures + JSON-Schema validation gate (see Tests below).

### 6. `product.rating` shape

Spec (`rating.json`):
```
required: ["value", "scale_min", "scale_max", "count"]
```

Today: `extract_rating()` at `class-wc-ai-storefront-ucp-product-translator.php:797-810` emits `{average, count}` only.

Fix:
- Rename `average` → `value`
- Add `scale_min: 1` (WC star ratings start at 1)
- Add `scale_max: 5`

~10 LOC, single-file change. Add a regression test in `UcpProductTranslatorTest`.

### 7. `product.seller` is non-spec at product level

Spec: `seller` is defined inline only on `variant.json` lines 149-165, optional, no required subfields (`name` + `links[]`). Not on product. No standalone `seller.json`.

Today: `product.seller: {name, country}` on every product. Fix: move to per-variant `seller: {name}` (drop `country` — not in spec; or move under `metadata` if we want to keep it). Drop product-level seller entirely.

Files:
- `class-wc-ai-storefront-ucp-product-translator.php:124-126`
- `class-wc-ai-storefront-ucp-rest-controller.php` `build_seller()`

### 8. `product.status`, `product.published_at`, `product.updated_at` are non-spec

Spec: not defined anywhere. `product.metadata` is the spec-blessed home for non-standard data.

Today: emitted at top level. Fix: move under `product.metadata.lifecycle: {status, published_at, updated_at}`.

Files:
- `class-wc-ai-storefront-ucp-product-translator.php:108-119`

### 9. Synthesized default variant has no `options[]` (prod_22)

For simple products with no options, this is fine. For misconfigured simple products with attribute terms, we currently route them to `metadata.attributes`. Document the rule explicitly. No code change.

### 10. `metadata.shipping` weight/dimensions as strings

WC returns these as strings; we forward verbatim. Spec doesn't constrain `metadata` shape, so this is not a violation. **Defer to a separate cleanup if ever.**

### 11. `pagination` in `/catalog/search` envelope

Spec: pagination is optional `$ref pagination.json#/$defs/response`. Today's emission matches. Verify against `pagination.json` field-name conventions during JSON-Schema validation.

## Plugin version

**0.12.0** — wire-format scope justifies a minor bump even though this is technically a compliance fix (we shipped non-conformant payloads under our declared `PROTOCOL_VERSION = 2026-04-08`). Spec-validating agents reading our responses today are already rejecting them; this PR fixes that. No coordination needed with public agent integrations because none exist that read our specific non-conformant keys.

## Tests

- Add JSON-Schema validation gate. Adopt the `release/2026-04-08` schemas in CI; run `opis/json-schema` (PHP) over fixture responses from `UcpProductTranslatorTest` and `UcpVariantTranslatorTest`. Catches all 10 gaps above and prevents regression.
- Update existing tests for the rename — `list_price` assertions become `price`, `compare_at_price` become `list_price`.
- New tests: per-variant `inputs[]` correlation in lookup, product.metadata.lifecycle shape, variant.seller (when present).

## Acceptance criteria

- [ ] All 6 critical gaps fixed
- [ ] All 5 important gaps fixed (or explicitly deferred with rationale)
- [ ] JSON-Schema validation gate added to CI; passes on all fixture responses
- [ ] Plugin version bumped to 0.12.0
- [ ] CHANGELOG documents the breaking shape changes

## Estimated effort

~250 LOC across 3 files, ~15 new/modified tests, ~half-day work + JSON-Schema validation setup.

### LINKED ISSUE #351: Performance: batch variation fetching for /catalog/search and /catalog/lookup
## Summary

`/catalog/search` and `/catalog/lookup` currently dispatch one `GET /wc/store/v1/products/{variation_id}` per variation per product. For a 20-product page where each variable product has N variations, this is `1 + (variable_count × N)` internal `rest_do_request` dispatches. The Store API supports `?parent_includes=` on the products collection endpoint — verified live: `GET /wc/store/v1/products?parent_includes=35` returns all 10 variations of product 35 in a single response.

Replacing the per-variation N+1 with one batched call per parent (or one global batched call for all parents in the page) eliminates the variation-expansion overhead.

## Verified live (2026-05-08, prod_35)

```bash
curl /wc/store/v1/products?include=91,92,93,94      # returns 0 — variations excluded by post_type filter
curl /wc/store/v1/products?parent=35                 # returns 0 — `parent` not the right param
curl /wc/store/v1/products?parent_includes=35        # returns 10 ← variations of product 35
```

`?parent_includes=<comma-separated parent IDs>` accepts multiple parents in a single call, returning all variations of any of them. Single batched call replaces N per-variation calls.

## Estimated performance gains

`rest_do_request()` cost on warm wp-env with object cache: ~15–40ms per dispatch. Cold: ~60–100ms.

| Scenario | Today (dispatches) | After batching | Latency saved (warm @ 25ms/call) |
|---|---|---|---|
| 20-product search, mostly simple (10% variable, avg 3 var) | 1 + 6 = **7** | 1 + 1 = **2** | ~125 ms |
| 20-product search, mixed (30% variable, avg 5 var) | 1 + 30 = **31** | 1 + 1 = **2** | ~725 ms |
| 20-product search, heavy variable (60% variable, avg 8 var) | 1 + 96 = **97** | 1 + 1 = **2** | ~2.4 s |
| 5-ID lookup, all simple | 5 = **5** | 5 = **5** | 0 |
| 5-ID lookup, all variable, avg 5 var | 5 + 25 = **30** | 5 + 1 = **6** | ~600 ms |
| 1-ID lookup, variable, 22 var (Leather Shoes) | 1 + 22 = **23** | 1 + 1 = **2** | ~525 ms |

Worst-case real-world (heavy apparel catalog, 20 products × 8 variations) saves multiple seconds per request — a meaningful improvement for any agent using the API in latency-sensitive flows.

### Cold-cache amplification

The savings scale with per-dispatch cost. A cold-cache request (right after `wp-env start` or after object cache flush) sees ~60-100ms per dispatch, multiplying the gain ~3x. The "heavy variable catalog" scenario goes from ~2.4s saved (warm) to ~7s saved (cold).

### 1. Add a batched fetch helper

`class-wc-ai-storefront-ucp-rest-controller.php`:

```php
/**
 * Fetch all variations for a set of parent product IDs in a single
 * Store API call. Returns a [parent_id => variations[]] map.
 *
 * @param int[] $parent_ids
 * @return array<int, array<int, array<string, mixed>>>
 */
private function fetch_variations_batched( array $parent_ids ): array {
    if ( empty( $parent_ids ) ) {
        return [];
    }
    $request = new WP_REST_Request( 'GET', '/wc/store/v1/products' );
    $request->set_query_params( [
        'parent_includes' => implode( ',', $parent_ids ),
        'per_page'        => 100, // Store API max
    ] );
    $response = rest_do_request( $request );
    // ... bin variations by `parent` field on each variation entry
}
```

### 2. Wire into `translate_products_for_search()` (line ~1041)

Replace the per-product `fetch_variations_for( $wc_product )` loop with:
- Pre-pass: collect all parent IDs that are variable products
- One batched call: `fetch_variations_batched( $parent_ids )`
- Per-product translate: pass the matching `variations[]` from the map (no fetch)

Same pattern for `translate_products_for_lookup()` (line ~1580).

### 3. Pagination edge case

Store API caps `per_page=100`. If a single page's combined variation count exceeds 100, split into multiple batched calls. Worst real case: 20 products × 8 variations = 160 — needs two batched calls. Still N+1 for the page-level slice but bounded at `ceil(total_vars / 100)` instead of `total_vars`.

### 4. Failure handling

Today's `fetch_variations_for()` returns a `skipped` count when individual variation fetches fail (logs `partial_variants` warning). The batched call either succeeds wholesale or fails wholesale — no per-variation failure mode. Test what happens when one parent's variations 404 (e.g., trashed parent with orphaned variations).

## Tradeoffs

**Pro:**
- Massive latency reduction for variable-heavy catalogs (verified scenarios above)
- Reduces DB query count proportionally (each `rest_do_request` runs full WC product hydration: post + meta + relationships + attributes)
- No spec/UCP shape change — purely internal optimization

**Con:**
- Larger response payload from Store API per call (~5KB per variation × 100 = ~500KB) — within process limits but worth monitoring
- Loses per-variation failure granularity (one variation 500ing kills the whole batch — investigate Store API behavior here)
- Adds complexity to the partial-variants warning path

## Acceptance criteria

- [ ] `fetch_variations_for()` (or equivalent) replaced with a batched implementation
- [ ] Both `/catalog/search` and `/catalog/lookup` use the batched path
- [ ] `partial_variants` warning still emitted correctly when some variations fail
- [ ] Tests:
  - [ ] Single-parent variable product (current behavior preserved)
  - [ ] Multi-parent search (20 products, 3 variable) — assert one batched dispatch
  - [ ] Pagination edge case (>100 combined variations) — assert two batched dispatches
  - [ ] Mixed-failure case — partial variants warning still fires
- [ ] Benchmark: measure latency before/after on a synthetic 20-product search response with 30% variable, 5 variations each. Document gains in PR description.

## Estimated effort

~150 LOC across 2 files, 4-6 new tests, ~3 hours including benchmarking.

## Plugin version

Patch (no shape change). 0.12.1 (after #349) or 0.13.1 (after #350).

### PR DESCRIPTION
**Pull request by pierorocca**
## Summary

Closes #351. Builds on PR #352 (closes #349) — targets the compliance branch as base; will rebase onto `main` after #352 merges. Ships as part of 0.12.0.

Replace per-variation `rest_do_request` N+1 fan-out with a single (or few) batched dispatch via `GET /wc/store/v1/products?type=variation&parent=<csv>&per_page=100`. Two params combine to surface variations through the collection endpoint:

- `type=variation` flips the underlying WP_Query's `post_type` from `'product'` (the default) to `'product_variation'` (per WC `StoreApi/Utilities/ProductQuery.php::prepare_objects_query()`). Without it, the route's default post_type filter excludes every variation regardless of the parent filter.
- `parent=<csv>` populates `post_parent__in` via WC's `wp_parse_id_list` sanitizer (per the same file + `StoreApi/Routes/V1/Products.php` parameter schema).

**Note (history):** an earlier revision of this PR mistakenly used `parent_includes=<csv>` — a parameter name that doesn't exist in WC's Store API. That dispatch was silently ignored by WP_Query and would have returned random catalog products in production. Tests passed because the `rest_do_request` stub recognized our fictional param. Caught + fixed before merge ([2d8397f](https://github.com/Automattic/woocommerce-ai-storefront/commit/2d8397f)) — verified live against a wp-env install: `?parent_includes=35` returned 18 random products with `parent: 0`; the corrected `?type=variation&parent=35` returns 9 actual variations of product 35.

## Performance gains

| Scenario | Before | After (warm) | Saved (warm) | Saved (cold) |
|---|---:|---:|---:|---:|
| 20 products, 30% variable, avg 5 vars (~30 var fetches) | 30 dispatches | 1 dispatch | ~700ms | ~2s+ |
| 20 products, 60% variable, avg 8 vars (~96 var fetches) | 96 dispatches | 1 dispatch | ~2.4s | ~7s |
| 5-ID lookup, all variable, avg 5 vars (~25 var fetches) | 30 dispatches | 6 dispatches | ~600ms | ~1.8s+ |
| 1-ID lookup, 22 variations (Leather Shoes) | 23 dispatches | 2 dispatches | ~525ms | ~1.5s+ |

(Estimates based on `rest_do_request` cost ~25ms warm / ~75ms cold-cache. Real numbers in the PR comment after benchmarking.)

## Five commits

| Commit | Subject |
|---|---|
| 1 | `perf(ucp): add fetch_variations_batched() helper (no callers yet)` |
| 2 | `perf(ucp): wire fetch_variations_batched into catalog/search` |
| 3 | `perf(ucp): wire fetch_variations_batched into catalog/lookup` |
| 4 | `refactor(ucp): remove fetch_variations_for() (was N+1 per-variation dispatch)` |
| 5 | `docs(changelog): add #351 perf entry under [0.12.0]` |

## Design

- **Single batched call vs per-parent batching:** chose single call. `rest_do_request` is in-process — no network failure mode that justifies per-parent isolation. The single-call approach reduces 15 dispatches (typical 30%-variable page) to 1.
- **Failure policy:** WP_Error or HTTP >= 400 on page 1 degrades every variable parent to `variations: [], skipped: total_declared` (degenerate fallback); page-N failures (N > 1) keep prior data and let per-parent `skipped = declared - returned` math compute residue. Each parent missing variations emits one `partial_variants` warning.
- **`per_page=100` cap:** paginates via short-page sentinel. For the rare page where combined-variation count > 100 we walk pages until short-page; bounded at `ceil(total_vars / 100)`.
- **Early-stop optimization:** the page loop tracks expected variation IDs in a flat set and breaks once every expected ID has been collected — for a parent with 200 variations capped at 50, walks 1 page instead of 2.
- **`MAX_VARIATIONS_PER_PRODUCT` cap:** preserved per-parent. Cap-overage contributes to `skipped` count so agents see `partial_variants` warning when their product overflows.
- **Scope gate:** variations inherit visibility from parent. Per-variation `is_product_syndicated()` was redundant and is now omitted (parent gate runs upstream).
- **`wc_ai_storefront_ucp_store_api_args` filter:** applies to the batched dispatch with the same `/wc/store/v1/products` endpoint string. The four dispatch-shape keys (`type`, `parent`, `page`, `per_page`) are restored after the filter so callbacks can't break the post_type targeting or the page-walk invariant.

## Tests

- New `UcpVariationBatchTest` (21 tests) covering the helper directly via `ReflectionClass` invocation: empty/all-simple guards, single-/multi-parent dispatch (asserts both `type=variation` AND `parent=<csv>`), >100-variation pagination, partial response, WP_Error / HTTP >= 400 degradation, MAX cap overage, early-stop pagination, source-order preservation, all-malformed-pointer regression, stdClass normalization, filter compatibility, non-array filter return.
- `UcpCatalogSearchTest` and `UcpCatalogLookupTest` stubs extended to recognize `?type=variation&parent=<csv>` on the collection endpoint and bin variations by `parent` field. Both params load-bearing — without `type=variation`, real WC defaults to `post_type='product'` and returns zero variations, so the stubs reject the call too.
- Three lookup tests rewritten to match the new dispatch contract (assert observable shape rather than per-ID dispatch counts).
- 1300 PHP tests passing.

## Test plan

- [x] `composer test` (full suite) — 1300/1300 passing
- [x] `vendor/bin/phpcs` clean
- [x] `vendor/bin/phpstan analyse` clean
- [x] Live verified: `?type=variation&parent=35&per_page=100` returns 9 actual variations of product 35 (vs old `?parent_includes=35` returning 18 random products)
- [ ] CI green on push (PHP 8.1/8.2/8.3/8.4, JS, lint, build, schema)

### CODE REVIEW

File: includes/ai-storefront/ucp-rest/class-wc-ai-storefront-ucp-rest-controller.php [hunk 1]
```diff
@@ -86,14 +86,16 @@ class WC_AI_Storefront_UCP_REST_Controller {
...
+	 * Upper bound on variations emitted per variable product.
+	 *
+	 * Pre-#351 this bounded a per-variation N+1 fan-out; post-#351
+	 * variations batch via `?parent_includes=` so the cap is now about
+	 * response payload size and translator workload rather than dispatch
+	 * count. Still meaningful: 50 covers typical color/size/pattern
+	 * combinations; products with more are rare and can surface via
+	 * `POST /catalog/lookup` with specific variation IDs if an agent
+	 * needs fidelity. Cap overage emits a `partial_variants` warning so
+	 * agents see when the cap kicks in.
```
[the remark this answers is not in the document]
Comment (pierorocca) [reply includes/ai-storefront/ucp-rest/class-wc-ai-storefront-ucp-rest-controller.php]: Fixed in 8e1b58d.

Right — the docblock had carried over wording from the pre-#351 era when the cap genuinely bounded fan-out (fewer expected IDs in the worklist meant fewer per-ID dispatches). Post-#351 the cap only bounds emission, because `parent_includes` returns ALL of a parent's variations regardless of our cap. The binner is what drops the over-cap tail.

Reworded the docblock to spell out the split:

```
What the cap DOES bound: emitted variant count, translator
workload, response payload size to the agent.

What the cap does NOT bound: Store API response size during
fetch, normalize_store_api_data() work on the discarded tail.
```

Also added a forward pointer to the early-stop optimization (your other finding, #3211063453) which DOES make the cap a real fetch-side bound for the typical case where WC returns variations in declared order.
